src/fpump.jl

```diff
@@ -0,0 +1,312 @@
+    js = MathProgBase.jac_structure(m.d)
+
+    jg = zeros(length(js[1]))
+    MathProgBase.eval_jac_g(m.d, jg, ones(m.num_var))
+
+    # Construct the data structure for our affine constraints
+    aff = Vector{Aff}(m.num_l_constr)
+    for i=1:m.num_l_constr
+        aff[i] = Aff()
+        aff[i].var_idx = []
+        aff[i].coeff = []
+        constr_expr = MathProgBase.constr_expr(m.d,i)
+        aff[i].rhs = constr_expr.args[3]
+        aff[i].sense = constr_expr.args[1]
+    end
+
+    # if linear constraint the derivative are the coeffs
+    idx = 1
+    lconstr2constr = Vector{Int64}()
+    constr2lconstr = Vector{Int64}()
+    c = 1
+    for i=1:m.num_constr
+        if m.isconstrlinear[i]
+            push!(lconstr2constr,i)
+            push!(constr2lconstr,c)
+            c+=1
+        else
+            push!(constr2lconstr,0)
+        end
+    end
+
+    for row in js[1]
+        if m.isconstrlinear[row]
+            col = js[2][idx]
+            aidx = constr2lconstr[row]
+            push!(aff[row].var_idx, col)
+            push!(aff[row].coeff, jg[idx])
+        end
+        idx += 1
+    end
+    return aff
+end
+
+"""
+    generate_mip(m, nlp_sol, aff, tabu_list)
+
+Generate a mip using the linear constraints (aff) of the original model
+Minimize the distance to nlp_sol and avoid using solutions inside the tabu list
+"""
+function generate_mip(m, nlp_sol, aff, tabu_list)
+    mip_model = Model(solver=m.mip_solver)
+    lb = [m.l_var; -1e6]
+    ub = [m.u_var; 1e6]
+    @variable(mip_model, lb[i] <= mx[i=1:m.num_var] <= ub[i])
+    for i=1:m.num_var
+        if m.var_type[i] == :Int
+            setcategory(mx[i], :Int)
+        elseif m.var_type[i] == :Bin
+            setcategory(mx[i], :Bin)
+        end
+    end
+    
+    for c in aff
+        if c.sense == :(>=)
+            @constraint(mip_model, sum(mx[c.var_idx[i]]*c.coeff[i] for i=1:length(c.var_idx)) >= c.rhs)
+        elseif c.sense == :(<=)
+            @constraint(mip_model, sum(mx[c.var_idx[i]]*c.coeff[i] for i=1:length(c.var_idx)) <= c.rhs)
+        elseif c.sense == :(==)
+            @constraint(mip_model, sum(mx[c.var_idx[i]]*c.coeff[i] for i=1:length(c.var_idx)) == c.rhs)
+        end
+    end
+
+    @variable(mip_model, mabsx[i=1:m.num_int_bin_var] >= 0)
+    for i=1:m.num_int_bin_var
+        vi = m.int2var_idx[i]
+        @constraint(mip_model, mabsx[i] >= mx[vi]-nlp_sol[vi])
+        @constraint(mip_model, mabsx[i] >= -mx[vi]+nlp_sol[vi])
+    end
+      
+    # How long is the tabu list 
+    num_sols = 0
+    for i=1:tabu_list.length
+        if !isnan(tabu_list.sols[i][1]) 
+            num_sols += 1
+        else 
+            break
+        end
+    end
+
+    # If there solutions in the tabu list => avoid them
+    if num_sols > 0
+        M = 100
+        @variable(mip_model, z1[j=1:m.num_int_bin_var,k=1:num_sols], Bin)
+        @variable(mip_model, z2[j=1:m.num_int_bin_var,k=1:num_sols], Bin)
+        v = tabu_list.sols
```

Comment:
This can be improved slightly to work over the set of values that appear in the solutions.  However, if this working as is, we can this leave it for now.

Reply:
That's true. This is the most basic simple implementation for now. Just to spot problems easier